### PR TITLE
python312Packages.biothings-client: init at 0.3.1 

### PR DESCRIPTION
**Pull request by rayhem**
## Description of changes

I'm super new to Nix packaging, but biothings-client is a dependency of MyGene (issue #344328). This is my first attempt at making a package rule, so reviews, suggestions, help, and guidance are very, _very_ appreciated 😄 

## Things done

I was able to build this locally with `nix shell` on Linux (NixOS itself), but, again, very new. I'm still learning how all the build stuff comes together.

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/development/python-modules/biothings-client/default.nix [hunk 1]
```diff
@@ -0,0 +1,34 @@
+{
+  lib,
+  buildPythonPackage,
+  fetchPypi,
+  setuptools,
+  requests,
+}:
+let
+  finalAttrs = {
```
Comment (dotlambda) [inline pkgs/development/python-modules/biothings-client/default.nix:9]:
```suggestion
buildPythonPackage rec {
```

File: pkgs/development/python-modules/biothings-client/default.nix [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+      hash = "sha256-yXK/LgK2+cx49/L7xe8CzFb+T4oq3LiAHskC9KtwEeY=";
+    };
+
+    build-system = [ setuptools ];
+    dependencies = [ requests ];
+    pythonImportsCheck = [ "biothings_client" ];
+
+    meta = {
+      changelog = "https://raw.githubusercontent.com/biothings/biothings_client.py/refs/heads/master/CHANGES.txt";
+      description = "biothings_client is an easy-to-use Python wrapper to access any Biothings.api -based backend service.";
```
Comment (dotlambda) [inline pkgs/development/python-modules/biothings-client/default.nix:26]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#meta-attributes

File: pkgs/development/python-modules/biothings-client/default.nix [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+      pname = "biothings_client";
+      hash = "sha256-yXK/LgK2+cx49/L7xe8CzFb+T4oq3LiAHskC9KtwEeY=";
+    };
+
+    build-system = [ setuptools ];
+    dependencies = [ requests ];
+    pythonImportsCheck = [ "biothings_client" ];
+
+    meta = {
+      changelog = "https://raw.githubusercontent.com/biothings/biothings_client.py/refs/heads/master/CHANGES.txt";
```
Comment (dotlambda) [inline pkgs/development/python-modules/biothings-client/default.nix:25]:
```suggestion
      changelog = "https://github.com/biothings/biothings_client.py/blob/v${version}/CHANGES.txt";
```

File: pkgs/development/python-modules/biothings-client/default.nix [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+      description = "biothings_client is an easy-to-use Python wrapper to access any Biothings.api -based backend service.";
+      homepage = "https://github.com/biothings/biothings_client.py";
+      license = lib.licenses.bsd3;
+      maintainers = with lib.maintainers; [ rayhem ];
+      mainProgram = "biothings_client";
```
Comment (dotlambda) [inline pkgs/development/python-modules/biothings-client/default.nix:30]: Does it really have a CLI?

**Review by dotlambda — DISMISSED** [review]
Please follow https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#commit-conventions.
You also need to add a line to `pkgs/top-level/python-packages.nix`.

File: pkgs/development/python-modules/biothings-client/default.nix (hunk 1 above)
Comment (rayhem) [reply pkgs/development/python-modules/biothings-client/default.nix]: Fixed

File: pkgs/development/python-modules/biothings-client/default.nix (hunk 3 above)
Comment (rayhem) [reply pkgs/development/python-modules/biothings-client/default.nix]: Fixed

File: pkgs/development/python-modules/biothings-client/default.nix (hunk 2 above)
Comment (rayhem) [reply pkgs/development/python-modules/biothings-client/default.nix]: Fixed

File: pkgs/development/python-modules/biothings-client/default.nix (hunk 4 above)
Comment (rayhem) [reply pkgs/development/python-modules/biothings-client/default.nix]: It does not 😄 I fumbled through this by looking at another package; I think I've fixed this per the top-level `README.md`

File: maintainers/maintainer-list.nix
Comment (dotlambda) [inline maintainers/maintainer-list.nix:1]: This should be done in a separate commit.

File: pkgs/development/python-modules/biothings-client/default.nix [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+  src = fetchPypi {
+    inherit version;
+    pname = "biothings_client";
+    hash = "sha256-yXK/LgK2+cx49/L7xe8CzFb+T4oq3LiAHskC9KtwEeY=";
+  };
+
+  build-system = [ setuptools ];
+  dependencies = [ requests ];
+  pythonImportsCheck = [ "biothings_client" ];
+
```
Comment (dotlambda) [inline pkgs/development/python-modules/biothings-client/default.nix:22]:
```suggestion
  nativeCheckInputs = [ pytestCheckHook ];
```

**Review by dotlambda — COMMENTED** [review]
Please squash and fix the commit message.

File: maintainers/maintainer-list.nix
Comment (rayhem) [reply maintainers/maintainer-list.nix:1]: I'm familiar with the git part but less-so the github part. Does this mean

1. Rebase my branch
2. Squash all the changes to one commit
3. Set the commit message to `biothings-client: init at 0.3.1` per the readme
4. Force-push it on top of my branch (since nothing else could depend on this branch yet)

? Will that leave the pull request in a valid state?

Comment (dotlambda) [reply maintainers/maintainer-list.nix:1]: No need to rebase before you squash. Mind the prefix in front of the commit message.

Comment (dotlambda) [reply maintainers/maintainer-list.nix:1]: And the maintainer commit should be `maintainers: add rayhem`.

Comment (dotlambda) [reply maintainers/maintainer-list.nix:1]: The prefix is still missing.

Comment (rayhem) [reply maintainers/maintainer-list.nix:1]: I'm sorry, I don't understand - my log shows two commits:

```
$ git log --oneline HEAD~2..HEAD
fa12bf898 (HEAD -> package-mygene) biothings-client: init at 0.3.1
c2862ede7 maintainers: add rayhem
```

Is `maintainers: add rayhem` incorrect? If `maintainers:` is the wrong prefix, I'm not sure where/how to put the right one.

Comment (dotlambda) [reply maintainers/maintainer-list.nix:1]: It should be `python312Packages.biothings-client: init at 0.3.1`.

Comment (rayhem) [reply maintainers/maintainer-list.nix:1]: Thanks for the help (and patience!) 🙂

File: pkgs/development/python-modules/biothings-client/default.nix [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+  setuptools,
+  pytestCheckHook,
+  requests,
+}:
+buildPythonPackage rec {
+  pname = "biothings-client";
+  version = "0.3.1";
+  pyproject = true;
+
+  src = fetchPypi {
```
Comment (natsukium) [inline pkgs/development/python-modules/biothings-client/default.nix:14]: PyPI tarball seems to have no tests. Could you fetch the source from GitHub instead?
https://github.com/NixOS/nixpkgs/pull/344340/checks?check_run_id=30805512453

Comment (rayhem) [reply pkgs/development/python-modules/biothings-client/default.nix:14]: I rigged this up and tested it locally, but it appears as though the large majority of their tests make network calls to pull in data over the network (which I gather is [not allowed in a derivation](https://discourse.nixos.org/t/using-dns-in-a-nix-derivation-temporary-failure-in-name-resolution/20543/4). Is there a standard way to handle that?

Comment (natsukium) [reply pkgs/development/python-modules/biothings-client/default.nix:14]: You can use the attributes `disabledTests` and `disabledTestPaths`.
https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md#using-pytestcheckhook-using-pytestcheckhook

Comment (rayhem) [reply pkgs/development/python-modules/biothings-client/default.nix]: I listed out the non-network tests that pass when I build locally with a comment indicating the rest require network access.

**Review by natsukium — APPROVED** [review]
LGTM, thanks.
Welcome to the nixpkgs!🎉
